### PR TITLE
Replace distance search with capped_distance in water bridge analysis

### PR DESCRIPTION
**Pull request by xiki-tempula**
Changes made in this Pull Request:
Changes the distance calculation to capped_distance in water bridge analysis.

PR Checklist
------------
 - [x] Tests?
 - [ ] Docs?
 - [x] CHANGELOG updated?
 - [ ] Issue raised/referenced?

### CODE REVIEW

**Comment by xiki-tempula** [comment]
I will add some test to bump the coverage up

**Comment by xiki-tempula** [comment]
@richardjgowers Hi Richard, I have bumped the coverage. I wonder if your mind does a code review? If you felt it is ok. I will modify the changelog.

File: package/MDAnalysis/analysis/hbonds/wbridge_analysis.py [hunk 1]
```diff
@@ -820,93 +821,108 @@ def _log_parameters(self):
...
+        self._s1_acceptors = []
 
-        if self.filter_first and self._s1:
+        self._s2_donors = []
+        self._s2_h_donors = {}
+        self._s2_donors_h = defaultdict(list)
+        self._s2_acceptors = []
+
+        self._s1 = self.u.select_atoms(self.selection1).ix
+        self._s2 = self.u.select_atoms(self.selection2).ix
```
Comment (richardjgowers) [inline package/MDAnalysis/analysis/hbonds/wbridge_analysis.py]: Here you're turning this selection into indices, but then only using these indices to slice `u.atoms` to get back the AtomGroup you used to have.  I'm not sure why this change was made?

Comment (xiki-tempula) [reply package/MDAnalysis/analysis/hbonds/wbridge_analysis.py]: As is mentioned in #2177, @orbeckst mentioned that it is better to store the relevant atom as atom index. This is a way of making internal representation consistent.
The more practical reason is that I need a double linked list which links the hydrogen bond donor heavy atom to the hydrogen as the bonded information is not always available. This is currently done using a dictionary. The key of the dictionary must be unmutable and thus, the atom index is chosen as a way of representing the atom. 
To avoid the confusion that in some part of the code, the atom group is represented as atom group and in another part of the code, atom group is represented as a list of indexes. I made the decision that all of the representation of atom group is done using atom index.

File: package/MDAnalysis/analysis/hbonds/wbridge_analysis.py [hunk 2]
```diff
@@ -987,29 +991,48 @@ def _prepare(self):
...
+        angles = np.rad2deg(
+            calc_angles(
+                self.u.atoms[tmp_donors].positions,
+                self.u.atoms[tmp_hydrogens].positions,
+                self.u.atoms[tmp_acceptors].positions,
+                box=self.box
+            )
+        )
+        hbond_indices = np.where(angles > self.angle)[0]
+        for index in hbond_indices:
```
[the remark this answers is not in the document]
Comment (xiki-tempula) [reply package/MDAnalysis/analysis/hbonds/wbridge_analysis.py]: Thank you for the suggestion. I will make it into a PR after this one is merged.

File: package/MDAnalysis/analysis/hbonds/wbridge_analysis.py [hunk 3]
```diff
@@ -1018,7 +1041,7 @@ def _single_frame(self):
 
         if self.update_selection:
             self._update_selection()
-        if self._s1 and self._s2:
```
[the remark this answers is not in the document]
Comment (xiki-tempula) [reply package/MDAnalysis/analysis/hbonds/wbridge_analysis.py]: Previously, the atom groups are represented as atom groups and thus, one can use `if self._s1 and self._s2:`.
However, in this PR, in order to implement the capped_distance. I need to construct a doubly-linked list, which could only be constructed with atom indexes. Thus, the _s1 and _s2 are list of atom indexes now and the `if self._s1 and self._s2:` no longer works.
Ideally, this should be done with bond manipulation #2382. I will do this in another PR.

**Comment by xiki-tempula** [comment]
@richardjgowers Thank you for the review. I guess the major problem in this PR is that I need to represent some atom group as a list of atoms indexes. To avoid the confusion that some atom groups are represented as indexes while others are represented as atom groups. I made that decision that atom groups should be represented as list of indexes, which changed some idioms. 
After I have studied the #2382, I will see if I can only use atom group and bonded information. However, the current idioms should work with both list of index and atom group and no api change or speed difference should be expected if I change the atom group representation back to atom group in another PR.

File: package/MDAnalysis/analysis/hbonds/wbridge_analysis.py (hunk 1 above)
Comment (richardjgowers) [reply package/MDAnalysis/analysis/hbonds/wbridge_analysis.py]: From what I can tell that issue is mostly about how the results are reported to users, internally you want to use whatever works best.  It doesn't make much sense to me to strip down an AtomGroup to indices only to reconstruct it later.

You can make a dictionary with `Atom` objects as keys.

But ok if this is how you want it and it works then we can go with this.

Comment (xiki-tempula) [reply package/MDAnalysis/analysis/hbonds/wbridge_analysis.py]: Thank you for the suggestion. I never know that atoms are unmutable, so when I need something unmutable, I turned to atom index. What is being manipulated here are all private attributes, so It is for use within the class only and is not meant to be shown to the user.

**Comment by xiki-tempula** [comment]
@richardjgowers  Yes, it has been benchmarked.
the order or water bridge versus the new and old implementation
```
0     1     2      3      4
1     6     21    52   123
1.3  12   40    95   205
```

File: package/MDAnalysis/analysis/hbonds/wbridge_analysis.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (xiki-tempula) [reply package/MDAnalysis/analysis/hbonds/wbridge_analysis.py]: After given it some thought, I'm not quite sure of the representation that you are talking about.
I will start with a list of indices (hydrogen bond donor heavy atom, hydrogen bond donor hydrogen, hydrogen bond acceptor), which is an array of rough size of (1000,3).
Then I will do the distance search, which gives the distance and shrink the number of indexes significantly. Thus, we have an array of (100,4), where the last column is distance.
Then I will do the angle search, which will also shrink the number of indexes and gives the final result of (10,5), where the last column is the angle.

To make a structured array, we need to know the array size in advance.
Do you suggest that I shall make a structured array of (1000,5) and then slice through it? to eventually get to (10,5).
Or is it better to reconstructed an array ar each step?
(1000,3) > (100,4) > (10,5)
Thank you.

**Comment by orbeckst** [comment]
Does this mean that the new implementation is much slower?

**Comment by xiki-tempula** [comment]
@orbeckst Sorry for the confusion. This is the amount of time which is required to run the same task. The new implementation runs twice as fast as the old one.

File: package/MDAnalysis/analysis/hbonds/wbridge_analysis.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (richardjgowers) [reply package/MDAnalysis/analysis/hbonds/wbridge_analysis.py]: You know the size of hbond_indices, so rather than have a loop iteration for each entry, you could create a structured array of size hbond_indices and fill that. Not too important though
